assets/backport/github/github.go

```diff
@@ -0,0 +1,398 @@
+// getBranch gets a branch.
+func (c *Client) getBranch(ctx context.Context, branchName string) (*go_github.Branch, error) {
+	branch, _, err := c.Client.Repositories.GetBranch(ctx, c.Organization, c.Repository, branchName, true)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return branch, nil
+}
+
+// createCommit creates a new commit.
+func (c *Client) createCommit(ctx context.Context, commitMessage string, tree *go_github.Tree, parent *go_github.Commit) (string, error) {
+	commit, _, err := c.Client.Git.CreateCommit(ctx, c.Organization, c.Repository, &go_github.Commit{
+		Message: &commitMessage,
+		Tree:    tree,
+		Parents: []*go_github.Commit{
+			parent,
+		},
+	})
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return commit.GetSHA(), nil
+}
+
+// getCommit gets a commit.
+func (c *Client) getCommit(ctx context.Context, sha string) (*go_github.Commit, error) {
+	commit, _, err := c.Client.Git.GetCommit(ctx,
+		c.Organization,
+		c.Repository,
+		sha)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return commit, nil
+}
+
+// merge merges a branch.
+func (c *Client) merge(ctx context.Context, base string, headCommitSHA string) (*go_github.Commit, error) {
+	merge, _, err := c.Client.Repositories.Merge(ctx, c.Organization, c.Repository, &go_github.RepositoryMergeRequest{
+		Base: &base,
+		Head: &headCommitSHA,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	mergeCommit, err := c.getCommit(ctx, merge.GetSHA())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return mergeCommit, nil
+}
+
+// GetBranchCommits gets commits on a branch.
+//
+// The only way to list commits for a branch is through RepositoriesService
+// and returns type RepositoryCommit which does not contain the commit
+// tree. To get the commit trees, GitService is used to get the commits (of
+// type Commit) that contain the commit tree.
+func (c *Client) GetBranchCommits(ctx context.Context, branchName string) ([]*go_github.Commit, error) {
+	// Getting RepositoryCommits.
+	repoCommits, err := c.getBranchCommits(ctx, branchName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Get the commits that are not on master. No commits will be returned if
+	// the pull request from the branch to backport was not squashed and merged
+	// or rebased and merged.
+	comparison, _, err := c.Client.Repositories.CompareCommits(ctx, c.Organization, c.Repository, "master", branchName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Getting Commits.
+	commits := []*go_github.Commit{}
+	for _, repoCommit := range repoCommits {
+		for _, diffCommit := range comparison.Commits {
+			if diffCommit.GetSHA() == repoCommit.GetSHA() {
+				commit, err := c.getCommit(ctx, repoCommit.GetSHA())
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+				if len(commit.Parents) != 1 {
+					return nil, trace.Errorf("merge commits are not supported.")
+				}
+				commits = append(commits, commit)
+			}
+		}
+	}
+	return commits, nil
+}
```

Comment:
I don't fully understand what this function is doing can you provide some more details?

I think the you're trying to accomplish is get the commits that should be backported right?

Reply:
What about something like this, since you are already getting the PR from the branch, can you get `HEAD` and `BASE` and just use that to get the commits you need to backport?

I wrote a little program to test it with this PR, it appears to work.

```go
package main

import (
    "context"
    "fmt"
    "log"

    "github.com/google/go-github/v41/github"
)

func main() {
    client := github.NewClient(nil)

    pr, _, err := client.PullRequests.Get(context.Background(),
        "gravitational",
        "teleport",
        9568)
    if err != nil {
        log.Fatal(err)
    }

    commits, _, err := client.Repositories.CompareCommits(
        context.Background(),
        "gravitational",
        "teleport",
        pr.GetBase().GetSHA(),
        pr.GetHead().GetSHA(),
        nil)
    if err != nil {
        log.Fatal(err)
    }

    for i, commit := range commits.Commits {
        fmt.Printf("--> %v: %v.\n", i, commit.GetSHA())
    }
}
```

Reply:
I thought of a simpler way to do this, I will update the PR.